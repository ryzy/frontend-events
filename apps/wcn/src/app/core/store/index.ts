import * as fromRouter from '@ngrx/router-store';
import { RouterReducerState } from '@ngrx/router-store';
import { Action, ActionReducer, ActionReducerMap, MetaReducer, RootStoreConfig } from '@ngrx/store';

import { appReducer, AppState } from './app/app-reducer';
import { AppRouterState } from './router/router';

/**
 * Store shape, for the root/core module
 * Feature stores are provided separately
 */
export interface AppRootState {
  app: AppState;
  router: RouterReducerState<AppRouterState>;
}

export const reducers: ActionReducerMap<AppRootState> = {
  app: appReducer as ActionReducer<AppState>,
  router: fromRouter.routerReducer as ActionReducer<RouterReducerState<AppRouterState>>,
};

export const rootStoreConfig: RootStoreConfig<AppRootState> = {
  metaReducers: [
    // logger,
  ],
  runtimeChecks: {
    // Cannot set these to true due to un-serializable ObjectID's from MongoDB...
    // strictActionImmutability: false,
    // strictStateImmutability: false,
  },
};

/**
 * Log all Store actions to console.
 * To use, uncomment it in `metaReducers` below
 */
/* istanbul ignore next */
export function logger(reducer: ActionReducer<AppRootState>): ActionReducer<AppRootState> {
  return function(state: AppRootState, action: Action): AppRootState {
    console.log('%cACTION ' + action.type, 'color: bisque; font-weight: bold', { action, stateBefore: state });
    return reducer(state, action);
  } as ActionReducer<AppRootState>;
}
