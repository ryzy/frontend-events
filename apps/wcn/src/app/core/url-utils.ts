import { Params } from '@angular/router';
import { ConferenceEvent, ConferenceEventRef } from '../event-base/model/conference-event';
import { EventTag } from '../event-base/model/event-tag';

import { EventsFilters } from '../event-base/model/events-filters';
import { AppSortInfo } from './model/entity';
import { AppRouterParams, AppRouterState } from './store/router/router';

export enum AppSectionUrls {
  Home = '/',
  EventsList = '/q',
  Event = '/ev',
}

/**
 * Make/clean params for router
 * We unset all empty variables, since we want nice/clean URLs
 */
export function makeParamsForRouter(params: Params, overrides: Params): Params {
  const newParams = { ...params, ...overrides };

  Object.keys(newParams).forEach((k) => {
    // make `true` shorter version of `1`
    if (true === newParams[k]) {
      newParams[k] = '1';
    }
    // remove all undefined's
    if (!newParams[k]) {
      delete newParams[k];
    }
  });

  return newParams;
}

export function makeEventsFiltersForRouter(filters: EventsFilters = {}): AppRouterParams {
  return <AppRouterParams>{
    where: filters.where || undefined,
    ws: filters.workshops ? '1' : undefined,
    fws: filters.freeWorkshops ? '1' : undefined,
    tags: filters.tags && filters.tags.length ? filters.tags.join(',') : undefined,
  };
}

export function getEventsFiltersFromRouter(state: AppRouterState): EventsFilters {
  return (
    state && {
      where: state.params.where || '',
      workshops: !!state.params.ws,
      freeWorkshops: !!state.params.fws,
      tags: state.params.tags ? state.params.tags.split(',') : undefined,
    }
  );
}

export const defaultSortInfo: AppSortInfo = {
  active: 'date',
  direction: 'asc',
};
export function makeSortInfoForRouter(sort?: AppSortInfo): Params {
  const params: Params = {
    s: undefined,
    sd: undefined,
  };

  if (sort && sort.active) {
    if (sort.active === defaultSortInfo.active && sort.direction === defaultSortInfo.direction) {
      // default sort info? return empty {}
      return params;
    } else if ('' === sort.direction) {
      // empty direction means: please remove this sorting, so we return immediately empty obj
      return params;
    } else if (sort.direction !== defaultSortInfo.direction) {
      // only generate direction for non-default direction
      params.sd = sort.direction;
    }

    params.s = sort.active;
  }

  return params;
}

/**
 * @see makeSortInfoForRouter
 */
export function getEventsSortInfoFromRouter(state: AppRouterState): AppSortInfo | undefined {
  return state && state.params.s
    ? <AppSortInfo>{
        active: state.params.s,
        direction: state.params.sd || defaultSortInfo.direction,
      }
    : undefined;
}

/**
 * Helper function to filter for events listing URLs
 */
export function filterEventsListingUrls(state: AppRouterState): boolean {
  return AppSectionUrls.Home === state.url || AppSectionUrls.EventsList === state.url;
}

/**
 * Get router link to an event
 */
export function getEventLink(
  ev: string | ConferenceEvent | ConferenceEventRef,
  baseUrl: AppSectionUrls = AppSectionUrls.Event,
): any[] {
  return [baseUrl, 'string' === typeof ev ? ev : ev.id];
}

/**
 * Get router link to tag(s)
 */
export function getTagLink(
  tag: string | string[] | EventTag | EventTag[],
  baseUrl: AppSectionUrls = AppSectionUrls.EventsList,
): any[] {
  let tags: string = '';
  if (Array.isArray(tag)) {
    tags = (tag as Array<string | EventTag>)
      .map((t: string | EventTag) => ('string' === typeof t ? t : t.id))
      .join(',');
  } else {
    tags = 'string' === typeof tag ? tag : tag.id;
  }

  return [baseUrl, { tags }];
}
