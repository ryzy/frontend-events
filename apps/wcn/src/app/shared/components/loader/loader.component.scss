.app-loader {
  position: absolute;
  z-index: 9;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  display: flex;
  flex-direction: column;
  align-items: center;
  background-color: rgba(255, 255, 255, 0.75);
  font-weight: 400;
  padding: 1rem 0;
}
/* Vertically centered version for <app-loader center=true> */
.app-loader.centered {
  justify-content: center;
}
/* Local (i.e. not absolute) version of loader */
.app-loader.local {
  position: relative;
}

@keyframes anim-rotate {
  100% {
    transform: rotate(360deg);
  }
}
@keyframes anim-dash {
  0% {
    stroke-dasharray: 1, 200;
    stroke-dashoffset: 0;
  }
  50% {
    stroke-dasharray: 89, 200;
    stroke-dashoffset: -35px;
  }
  100% {
    stroke-dasharray: 89, 200;
    stroke-dashoffset: -124px;
  }
}

.app-loader .spinner {
  position: relative;
  width: 5rem;
}
.app-loader svg {
  animation: anim-rotate 2s linear infinite;
  transform-origin: center;
}
.app-loader circle {
  stroke: #d81b60;
  stroke-dasharray: 1, 200;
  stroke-dashoffset: 0;
  animation: anim-dash 1.5s ease-in-out infinite;
}
