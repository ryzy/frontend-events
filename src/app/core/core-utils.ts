/**
 * Prevent re-import of core modules
 *
 * @url https://angular.io/guide/styleguide#style-04-12
 *
 * Usage:
 * ```
 * export class MyModule {
 *   constructor(@Optional() @SkipSelf() parentModule: MyModule) {
 *     throwIfAlreadyLoaded(parentModule, 'MyModule');
 *   }
 * }
 * ```
 */
export function throwIfAlreadyLoaded(parentModule: any, moduleName: string): void {
  if (parentModule) {
    throw new Error(`${moduleName} has already been loaded. Import Core modules in the AppModule only.`);
  }
}

/**
 * Returns true if code is executed in test (i.e. Karma) context.
 */
export function isUnitTestContext(): boolean {
  return !!(window as any).__karma__;
}

// tslint:disable:no-bitwise
/**
 * Generate UUID string
 */
export function uuid(length?: number): string {
  const str = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c: string) => {
    const r: number = (Math.random() * 16) | 0;
    const v: number = c === 'x' ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });

  return length ? str.substr(0, length) : str;
}
// tslint:enable:no-bitwise

/**
 * Get random number from given range
 */
export function randomRange(min: number = 0, max: number = 999): number {
  return Math.floor(Math.random() * (max - min)) + min;
}
