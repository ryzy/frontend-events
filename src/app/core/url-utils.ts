import { Params } from '@angular/router';

import { EventsFilters } from '../event-base/model/events-filters';
import { AppSortInfo } from './model/entity';
import { AppRouterState } from './store/router/router';

export enum AppSectionUrls {
  Home = '/',
  EventsList = '/q',
}

/**
 * Make/clean params for router
 * We unset all empty variables, since we want nice/clean URLs
 */
export function makeParamsForRouter(params: Params, overrides: Params): Params {
  const newParams = { ...params, ...overrides };

  Object.keys(newParams).forEach((k) => {
    // make `true` shorter version of `1`
    if (true === newParams[k]) {
      newParams[k] = '1';
    }
    // remove all undefined's
    if (!newParams[k]) {
      delete newParams[k];
    }
  });

  return newParams;
}

export function makeEventsFiltersForRouter(filters: EventsFilters = {}): Params {
  return {
    where: filters.where || undefined,
    ws: filters.workshops ? '1' : undefined,
    fws: filters.freeWorkshops ? '1' : undefined,
  };
}

export function getEventsFiltersFromRouter(state: AppRouterState): EventsFilters {
  return (
    state && {
      where: state.params.where || '',
      workshops: !!state.params.ws,
      freeWorkshops: !!state.params.fws,
    }
  );
}

export const defaultSortInfo: AppSortInfo = {
  active: 'date',
  direction: 'asc',
};
export function makeSortInfoForRouter(sort?: AppSortInfo): Params {
  const params: Params = {
    s: undefined,
    sd: undefined,
  };

  if (sort && sort.active) {
    if (sort.active === defaultSortInfo.active && sort.direction === defaultSortInfo.direction) {
      // default sort info? return empty {}
      return params;
    } else if ('' === sort.direction) {
      // empty direction means: please remove this sorting, so we return immediately empty obj
      return params;
    } else if (sort.direction !== defaultSortInfo.direction) {
      // only generate direction for non-default direction
      params.sd = sort.direction;
    }

    params.s = sort.active;
  }

  return params;
}

/**
 * @see makeSortInfoForRouter
 */
export function getEventsSortInfoFromRouter(state: AppRouterState): AppSortInfo | undefined {
  return state && state.params.s
    ? <AppSortInfo>{
        active: state.params.s,
        direction: state.params.sd || defaultSortInfo.direction,
      }
    : undefined;
}
