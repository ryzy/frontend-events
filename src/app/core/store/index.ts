import { Action, ActionReducer, ActionReducerMap, MetaReducer } from '@ngrx/store';
import * as fromRouter from '@ngrx/router-store';

import { appReducer, AppState } from './app/app-reducer';
import { AppRouterState } from './router/router';

/**
 * Store shape, for the root/core module
 * Feature stores are provided separately
 */
export interface AppRootState {
  app: AppState;
  router: fromRouter.RouterReducerState<AppRouterState>;
}

export const reducers: ActionReducerMap<AppRootState> = {
  app: appReducer as ActionReducer<AppState>,
  router: fromRouter.routerReducer as ActionReducer<fromRouter.RouterReducerState<AppRouterState>>,
};

/**
 * Log all Store actions to console.
 * To use, uncomment it in `metaReducers` below
 */
/* istanbul ignore next */
export function logger(reducer: ActionReducer<AppRootState>): ActionReducer<AppRootState> {
  return function(state: AppRootState, action: Action): AppRootState {
    console.log('%cACTION ' + action.type, 'color: blue; font-weight: bold', { action, stateBefore: state });
    return reducer(state, action);
  } as ActionReducer<AppRootState>;
}

/* istanbul ignore next */
export const metaReducers: MetaReducer<AppRootState>[] = [
  // logger
];
