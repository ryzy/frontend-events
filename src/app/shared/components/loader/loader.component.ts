import { Component, ChangeDetectionStrategy, Input, ViewEncapsulation } from '@angular/core';

@Component({
  selector: 'app-loader',
  host: {
    class: 'app-loader',
    '[class.centered]': 'centered',
  },
  template: `
    <div class="spinner">
      <svg class="circular" viewBox="25 25 50 50">
        <circle class="path" cx="50" cy="50" r="20" fill="none" stroke-width="2" stroke-miterlimit="10"/>
      </svg>
    </div>
    <h4 *ngIf="message">{{message}}</h4>
    <ng-content></ng-content>
  `,
  styleUrls: ['./loader.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  encapsulation: ViewEncapsulation.None,
})
export class LoaderComponent {
  @Input()
  public message: string = 'Loading...';

  @Input()
  public centered: boolean = false;
}
