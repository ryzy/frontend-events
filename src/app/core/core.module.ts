import { NgModule, Optional, SkipSelf } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { ServiceWorkerModule } from '@angular/service-worker';
import { EffectsModule } from '@ngrx/effects';
import { RouterStateSerializer, StoreRouterConnectingModule } from '@ngrx/router-store';
import { StoreModule } from '@ngrx/store';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';
import { AngularFireModule } from 'angularfire2';
import { AngularFireAuthModule } from 'angularfire2/auth';
import { AngularFirestoreModule } from 'angularfire2/firestore';

import { environment } from '../../environments/environment';
import { throwIfAlreadyLoaded } from './core-utils';
import { reducers, metaReducers } from './store/index';
import { RouterEffects } from './store/router-effects';
import { AppRouterStateSerializer } from './store/router-state-serializer';

@NgModule({
  imports: [
    BrowserModule,
    BrowserAnimationsModule,

    // @ngrx store
    StoreModule.forRoot(reducers, { metaReducers }),
    StoreRouterConnectingModule,
    /* istanbul ignore next */
    !environment.production ? StoreDevtoolsModule.instrument({ name: 'what-conference-next.com' }) : [],
    EffectsModule.forRoot([RouterEffects]),

    // Firebase setup
    AngularFireModule.initializeApp(environment.firebase),
    AngularFireAuthModule,
    AngularFirestoreModule,

    // SW
    ServiceWorkerModule.register('/ngsw-worker.js', { enabled: environment.production }),
  ],
  providers: [
    { provide: RouterStateSerializer, useClass: AppRouterStateSerializer },
  ]
})
export class CoreModule {
  public constructor(@Optional() @SkipSelf() parentModule?: CoreModule) {
    throwIfAlreadyLoaded(parentModule, 'CoreModule');
  }
}
