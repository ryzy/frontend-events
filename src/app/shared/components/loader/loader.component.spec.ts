import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { LoaderComponent } from './loader.component';

describe('LoaderComponent', () => {
  let component: LoaderComponent;
  let fixture: ComponentFixture<LoaderComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [LoaderComponent],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(LoaderComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    component.local = true; // just so it's not full screen on Karma results
    component.message = 'Loader Test';
    fixture.detectChanges();
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });
});
