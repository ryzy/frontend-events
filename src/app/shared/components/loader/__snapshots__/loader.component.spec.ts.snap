// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoaderComponent should create 1`] = `
<app-loader
  centered="false"
  local={[Function Boolean]}
  message={[Function String]}
>
  <div
    class="spinner"
  >
    <svg
      class="circular"
      viewBox="25 25 50 50"
    >
      <circle
        class="path"
        cx="50"
        cy="50"
        fill="none"
        r="20"
        stroke-miterlimit="10"
        stroke-width="2"
      />
    </svg>
  </div><h4>
    Loader Test
  </h4>
</app-loader>
`;
